pkg/virt-controller/watch/vm.go

```diff
@@ -2025,6 +1921,130 @@ func (c *VMController) processFailureCondition(vm *virtv1.VirtualMachine, vmi *v
+		stateChange := vm.Status.StateChangeRequests[0]
+		switch stateChange.Action {
+		case virtv1.StopRequest:
+			if vmi == nil {
+				// because either the VM or VMI informers can trigger processing here
+				// double check the state of the cluster before taking action
+				_, err := c.clientset.VirtualMachineInstance(vm.ObjectMeta.Namespace).Get(vm.GetName(), &v1.GetOptions{})
+				if err != nil && errors.IsNotFound(err) {
+					// If there's no VMI, then the VMI was stopped, and the stopRequest can be cleared
+					log.Log.Object(vm).V(4).Infof("No VMI. Clearing stop request")
+					clearChangeRequest = true
+				}
+			} else {
+				if stateChange.UID == nil {
+					// It never makes sense to have a request to stop a VMI that doesn't
+					// have a UUID associated with it. This shouldn't be possible -- but if
+					// it occurs, clear the stopRequest because it can't be acted upon
+					log.Log.Object(vm).Errorf("Stop Request has no UID.")
+					clearChangeRequest = true
+				} else if *stateChange.UID != vmi.UID {
+					// If there is a VMI, but the UID doesn't match, then it
+					// must have been previously stopped, so the stopRequest can be cleared
+					log.Log.Object(vm).V(4).Infof("VMI's UID doesn't match. clearing stop request")
+					clearChangeRequest = true
+				}
+			}
+		case virtv1.StartRequest:
+			// If the current VMI is running, then it has been started.
+			if vmi != nil {
+				log.Log.Object(vm).V(4).Infof("VMI exists. clearing start request")
+				clearChangeRequest = true
+			}
+		}
+	}
+
+	return clearChangeRequest
+}
+
+func (c *VMController) trimDoneVolumeRequests(vm *virtv1.VirtualMachine) {
+	if len(vm.Status.VolumeRequests) > 0 {
+		volumeMap := make(map[string]virtv1.Volume)
+		diskMap := make(map[string]virtv1.Disk)
+
+		for _, volume := range vm.Spec.Template.Spec.Volumes {
+			volumeMap[volume.Name] = volume
+		}
+		for _, disk := range vm.Spec.Template.Spec.Domain.Devices.Disks {
+			diskMap[disk.Name] = disk
+		}
+
+		tmpVolRequests := vm.Status.VolumeRequests[:0]
+		for _, request := range vm.Status.VolumeRequests {
+
+			var added bool
+			var volName string
+
+			removeRequest := false
+
+			if request.AddVolumeOptions != nil {
+				volName = request.AddVolumeOptions.Name
+				added = true
+			} else if request.RemoveVolumeOptions != nil {
+				volName = request.RemoveVolumeOptions.Name
+				added = false
+			}
+
+			_, volExists := volumeMap[volName]
+			_, diskExists := diskMap[volName]
+
+			if added && volExists && diskExists {
+				removeRequest = true
+			} else if !added && !volExists && !diskExists {
+				removeRequest = true
+			}
+
+			if !removeRequest {
+				tmpVolRequests = append(tmpVolRequests, request)
+			}
+		}
+		vm.Status.VolumeRequests = tmpVolRequests
+	}
+}
+
+func (c *VMController) syncChangeRequest(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance, key string, dataVolumes []*cdiv1.DataVolume) (syncError, error) {
+	var syncErr syncError
```

Comment:
This is not entirely true since, for example, `syncErr = c.startStop(vm, vmi)` can return nil. Also, `virtControllerVMWorkQueueTracer.StepTrace(key, "sync", trace.Field{Key: "VM Name", Value: vm.Name})` at the end of the function needs to be called regardless if `syncErr` is nil or not.

I think that returning at the middle of the function in this case might cause more confusion than clarity. I think that, in general, a single return point (except for the first one that is trivial to read) is a good thing, especially if there are operations you want to perform regardless of the return value.

Reply:
Right, good point about `startStop` returning nil.

> especially if there are operations you want to perform regardless of the return value.

That's sometimes a sign that a function is not cohesive enough, or is doing several things which can/should be further broken apart. Nevertheless, this is sometimes easier to say than actually do :-)

Reply:
Btw, we could `defer virtControllerVMWorkQueueTracer.StepTrace(key, "sync", trace.Field{Key: "VM Name", Value: vm.Name})` to ensure that it's always called on the way out.

Reply:
> That's sometimes a sign that a function is not cohesive enough

Well, sometimes :) in general a "cohesive task" is not something that can be defined objectively. One task can be performed of several operations, and a function that serves that task is cohesive. It's like "unit test" - we need a test for every unit - but a unit is not a well-defined term. IMO a function that also performs loggings of its actions (by calling another function) is indeed cohesive.

> Btw, we could defer virtControllerVMWorkQueueTracer.StepTrace(key, "sync", trace.Field{Key: "VM Name", Value: vm.Name}) to ensure that it's always called on the way out.

As I said above - I see benefit in having a single return point from a function. Such code tends to me clearer and less error-prone and easier to debug (and even in terms of performance - compiler tends to be able to optimize such code much better).

If you insist I can change that, but I'm not entirely sure it'll make things better

Reply:
I definitely don't insist :-)
That's perfectly fine as it is now.